media-proxy/src/mesh/conn_rdma.cc

```diff
@@ -0,0 +1,359 @@
+ * 
+ * @param capacity The number of buffers to allocate within continuous virtual memory.
+ * @param trx_sz The size of each buffer (aligned to the page size).
+ * @return Result::success on successful initialization, or an error result if arguments are invalid 
+ *         or memory allocation fails.
+ */
+Result Rdma::init_queue_with_elements(size_t capacity, size_t trx_sz)
+{
+    if (trx_sz == 0 || trx_sz > MAX_BUFFER_SIZE) {
+        log::error("Invalid transfer size provided for buffer allocation")("trx_sz", trx_sz);
+        return Result::error_bad_argument;
+    }
+
+    std::lock_guard<std::mutex> lock(queue_mutex);
+
+    // Check if already initialized
+    if (!buffer_queue.empty()) {
+        log::error("Buffer queue already initialized");
+        return Result::error_already_initialized;
+    }
+
+    // Calculate total memory size needed and align to page size
+    size_t aligned_trx_sz = ((trx_sz + PAGE_SIZE - 1) / PAGE_SIZE) * PAGE_SIZE;
+    size_t total_size = capacity * aligned_trx_sz;
+
+    // Allocate a single contiguous memory block
+    void *memory_block = std::aligned_alloc(PAGE_SIZE, total_size);
+    if (!memory_block) {
+        log::error("Rdma failed to allocate a single memory block")("total_size", total_size);
+        return Result::error_out_of_memory;
+    }
+
+    // Zero-initialize the entire memory block
+    std::memset(memory_block, 0, total_size);
+
+    // Divide the memory block into individual buffers
+    char *base_ptr = static_cast<char *>(memory_block);
+    for (size_t i = 0; i < capacity; ++i) {
+        void *buf = base_ptr + i * aligned_trx_sz;
+        buffer_queue.push(buf);
+    }
+
+    // Store the base memory block for cleanup
+    buffer_block = memory_block;
+
+    return Result::success;
+}
+
+// Cleanup the buffer queue
+void Rdma::cleanup_queue()
+{
+    if (buffer_block) {
+        // Free the entire memory block
+        std::free(buffer_block);
+        buffer_block = nullptr;
+    }
+    while (!buffer_queue.empty()) {
+        buffer_queue.pop();
+    }
+}
+
+/**
+ * @brief Configures the RDMA connection with the provided parameters.
+ * 
+ * Sets up the RDMA transfer size, connection kind, endpoint configuration,
+ * and queue size based on the input request. Ensures that the kind-direction
+ * pairing is valid (e.g., receiver with RX direction). Updates the connection
+ * state to "configured" upon success.
+ * 
+ * @param ctx The operation context.
+ * @param request The connection parameters including addresses and RDMA arguments.
+ * @param dev_port The device port identifier.
+ * @param dev_handle Reference to the RDMA device handle.
+ * @param kind The type of connection (receiver or transmitter).
+ * @param dir The data transfer direction (RX or TX).
+ * @return Result::success on successful configuration, or an error result if arguments are invalid.
+ */
+Result Rdma::configure(context::Context& ctx, const mcm_conn_param& request,
```

Comment:
is `dev_port` used in any way? if yes, why is it outside mcm_conn_param, if no, why does it exist?

Reply:
It's in place to allow other type of device identification than pure ip address, this proves to be problematic in some scenarios, verbs sometimes tends to treat passed ip address as suggestion not requirement for choosing a device

Reply:
It is great if you can use PCI port to select the device. My point is that I can't see that happening here. `dev_port` is taken as parameter to `configure()`, but it is never used.

Reply:
fixed in https://github.com/OpenVisualCloud/Media-Communications-Mesh/pull/286/commits/c49f5eba0ccf59ae374fbfcf4cdfd4888363324d